Add support for WASM targets

Luthaf opened the pull request:
This PR adds support for `Monotonic` on both WASM targets: `wasm32-unknown-unknown` (web) and `wasm32-wasi` (wasi). On the web, it uses [window.performance.now()](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now); and on wasi it uses [clock_time_get](https://docs.rs/wasi/0.10.2+wasi-snapshot-preview1/wasi/fn.clock_time_get.html) with a monotonic clock.

I did not ran the tests on these targets, but I'll try to do so in the next few days.

tobz commented:
At first glance, this looks reasonable to me.  I'll try to do a full review on it in the next few days.  The cfg conditionals are starting to get a bit hairy and I want to make sure things are 👌🏻  😅

Luthaf commented:
> The cfg conditionals are starting to get a bit hairy and I want to make sure things are 👌🏻 😅

Yeah, wasm being available on both web and wasi do not help here ...

tobz commented on src/monotonic.rs:
```diff
@@ -76,6 +78,46 @@ impl Monotonic {
 }
 
+#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+impl Monotonic {
+    pub fn new() -> Monotonic {
+        Monotonic {
+            factor: 0,
+        }
+    }
+}
```
You should drop this impl, and the one below, and then remove the addition you made on line 13.

tobz commented on src/monotonic.rs:
```diff
@@ -76,6 +78,46 @@ impl Monotonic {
+#[cfg(all(target_arch = "wasm32", target_os = "unknown"))]
+impl Monotonic {
+    pub fn now(&self) -> u64 {
+        let now = web_sys::window()
+                    .expect("should have a Window")
+                    .performance()
+                    .expect("should have a Performance")
+                    .now();
```
I'm not an expert on WASM, but is the implication of the target arch/target OS combination here that this should only ever run in the browser?  And is it also a given that `window.performance` should always be available?  I feel like I remember `window.performance.now()` being neutered back when Rowhammer was announced... either by limiting its availability or reducing its resolution or both?

Luthaf replied:
> but is the implication of the target arch/target OS combination here that this should only ever run in the browser?

Unfortunately no, I've been a bit optimistic here. `wasm32-unkown-unkown` can also run on other WASM VM, in particular node.js, wasmer, etc. While node also offer [performance.now](https://nodejs.org/api/perf_hooks.html#perf_hooks_performance_now), there is not generic way (contrarily to wasi) to get such functionality in any VM. However, I think that most of the non-browser users of wasm are using wasi, and as such only supporting browsers in this function could be fine (maybe with an improved panic message pointing users to wasi?)

If we are running in browser, `window.performance.now` will always be available (https://developer.mozilla.org/en-US/docs/Web/API/Performance/now). Its precision might be very bad (for example Firefox defaults to a precision of 1ms, and can be configured to use a precision of 100ms), but the function will be there. I'm personally fine with limited precision: I'd rather have something working/compiling from the same codebase with limited precision than nothing.

tobz replied:
Yeah, all understandable.

I think what we probably need to do here is add a blurb to the module-level docs that talk about WASM support and briefly explain the shortcomings with regards to accuracy.  While the performance of taking the measurements themselves is the primary goal of the library, we implicitly provide nanosecond resolution _and_ accuracy with measurements on our more traditional platforms.  Going from nanosecond accuracy to millisecond or centisecond accuracy would be.... a huge jump. :)

Luthaf replied:
Adding documentation on this subject sounds good, I'll give it a go when I find some time this week!

Luthaf commented:
I cleaned up the code as suggested in the review, and added support to run the tests.

### running tests for wasm32-wasi

```bash
# install wasmtime
curl https://wasmtime.dev/install.sh -sSf | bash
export CARGO_TARGET_WASM32_WASI_RUNNER="wasmtime run"
cargo test --target wasm32-wasi
```

### running tests for wasm32-unknown-unknown

```bash
# install wasm-pack
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh 
# run tests on firefox
wasm-pack test --firefox --headless
# run tests on chrome
wasm-pack test --chrome --headless

# run tests on node (require commenting out the configure_wasm_tests module in lib.rs)
# this currently fails with ReferenceError: Window is not defined due to 
wasm-pack test --node
```

tobz commented on src/upkeep.rs:
```diff
@@ -155,6 +155,7 @@ mod tests {
     use std::time::Duration;
 
     #[test]
+    #[cfg_attr(target_arch = "wasm32", ignore)] // WASM is single threaded
```
For my own curiosity: does `std::thread::spawn` not work at all on WASM?

Luthaf replied:
At least currently, yes. There are proposals to add threading to WASM, both in browsers and WASI, but they are still in early stage.

Luthaf commented:
I've added some documentation on accuracy, let me know if there is a better way to do express these limitations

tobz commented:
@Luthaf Changes look good to me, just need to run `cargo fmt` and this should be set to merge.

tobz commented:
@Luthaf Sorry for the delay here.  `quanta@v0.9.0` is now released and contains your PR.

Thanks again for your contribution! 🌮

Luthaf commented:
thanks a lot for the release!
